sycl/include/CL/sycl/intel/group_algorithm.hpp

```diff
@@ -0,0 +1,635 @@
+//==----------- group_algorithm.hpp --- SYCL group algorithm----------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+#include <CL/__spirv/spirv_ops.hpp>
+#include <CL/__spirv/spirv_types.hpp>
+#include <CL/__spirv/spirv_vars.hpp>
+#include <CL/sycl/detail/spirv.hpp>
+#include <CL/sycl/detail/type_traits.hpp>
+#include <CL/sycl/group.hpp>
+#include <CL/sycl/intel/functional.hpp>
+
+#ifndef __DISABLE_SYCL_INTEL_GROUP_ALGORITHMS__
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace detail {
+
+template <typename Group> size_t get_local_linear_range(Group g);
+template <> size_t get_local_linear_range<group<1>>(group<1> g) {
+  return g.get_local_range(0);
+}
+template <> size_t get_local_linear_range<group<2>>(group<2> g) {
+  return g.get_local_range(0) * g.get_local_range(1);
+}
+template <> size_t get_local_linear_range<group<3>>(group<3> g) {
+  return g.get_local_range(0) * g.get_local_range(1) * g.get_local_range(2);
+}
+
+template <int Dimensions>
+id<Dimensions> linear_id_to_id(range<Dimensions>, size_t i);
+template <> id<1> linear_id_to_id(range<1> r, size_t i) { return id<1>(i); }
+template <> id<2> linear_id_to_id(range<2> r, size_t i) {
+  id<2> result;
+  result[0] = i / r[1];
+  result[1] = i % r[1];
+  return result;
+}
+template <> id<3> linear_id_to_id(range<3> r, size_t i) {
+  id<3> result;
+  result[0] = i / (r[1] * r[2]);
+  result[1] = (i % (r[1] * r[2])) / r[2];
+  result[2] = i % r[2];
+  return result;
+}
+
+template <typename T> struct is_group : std::false_type {};
+
+template <int Dimensions>
+struct is_group<group<Dimensions>> : std::true_type {};
+
+template <typename T, class BinaryOperation> struct identity {};
+
+template <typename T, typename V> struct identity<T, intel::plus<V>> {
+  static constexpr T value = 0;
+};
+
+template <typename T, typename V> struct identity<T, intel::minimum<V>> {
+  static constexpr T value = std::numeric_limits<T>::max();
+};
+
+template <typename T, typename V> struct identity<T, intel::maximum<V>> {
+  static constexpr T value = std::numeric_limits<T>::lowest();
+};
+
+template <typename Group, typename Ptr, class Function>
+Function for_each(Group g, Ptr first, Ptr last, Function f) {
+#ifdef __SYCL_DEVICE_ONLY__
+  nd_item<Group::dimensions> it =
+      ::sycl::detail::Builder::getNDItem<Group::dimensions>();
+  ptrdiff_t offset = it.get_local_linear_id();
+  ptrdiff_t stride = detail::get_local_linear_range(g);
+  for (Ptr p = first + offset; p < last; p += stride) {
+    f(*p);
+  }
+  return f;
+#else
+  throw runtime_error("Group algorithms are not supported on host device.");
```

Comment:
This is intended to be `std::runtime_error` and not `sycl::runtime_error`, right?

Note: error code is mandatory field fro `sycl::` exceptions since #1150

Reply:
I had copied this from an older version of the sub-group host implementation.  All the `runtime_error`s there have been updated with a `PI_INVALID_DEVICE` argument, so I'll make the same change here.

Reply:
Fixed in https://github.com/intel/llvm/pull/1236/commits/3f4e8e77596f1e2da8553f2c0c0bb51ce501e727.